### PR TITLE
refactor speedy.Compiler.freeVars to be functional

### PR DESCRIPTION
**Pull request by garyverhaegen-da**
I was trying to track down what looks like a new source of flakiness in the ANF work (#6440), which manifested in a stack overflow on the recursion of the `go` function in speedy.Compiler. I spent a bit of time refactoring this to be functional in order to then trampoline it, only to realize afterwards that I had worked on `freeVars/go` whereas the one blowing up the stack was `validate/go`.

Obviously, it did nothing to reduce the rate of flakiness on the stack overflow issue I was trying to address, as it's a separate function.

So there isn't really a good reason for this change, except I had done it and I do believe the code looks marginally nicer, so might as well submit it.

CHANGELOG_BEGIN
CHANGELOG_END

### CODE REVIEW

**Comment by cocreature** [comment]
I’ll leave it to someone else to judge whether we want to merge this (code looks reasonable to me) but I can’t help but point to the Ben’s excellent blog post on efficient variable traversals in Haskell https://www.haskell.org/ghc/blog/20190728-free-variable-traversals.html (yes I know Haskell performance characteristics are different so it doesn’t necessarily apply here but it’s still interesting)

File: daml-lf/interpreter/src/main/scala/com/digitalasset/daml/lf/speedy/Compiler.scala [hunk 1]
```diff
@@ -1125,67 +1125,50 @@ private[lf] final case class Compiler(
     * The returned free variables are de bruijn indices
     * adjusted to the stack of the caller. */
   def freeVars(expr: SExpr, initiallyBound: Int): Set[Int] = {
```
Comment (remyhaemmerle-da) [inline daml-lf/interpreter/src/main/scala/com/digitalasset/daml/lf/speedy/Compiler.scala:1127]:
```suggestion
  def freeVars(expr: SExpr, initiallyBound: Int, free: Set[Int]): Set[Int] = {
```

File: daml-lf/interpreter/src/main/scala/com/digitalasset/daml/lf/speedy/Compiler.scala [hunk 2]
```diff
@@ -1125,67 +1125,50 @@ private[lf] final case class Compiler(
...
+    def go(expr: SExpr, bound: Int): Set[Int] =
       expr match {
         case SEVar(i) =>
-          if (i > bound)
-            free += i - bound /* adjust to caller's environment */
-        case _: SEVal => ()
-        case _: SEBuiltin => ()
-        case _: SEValue => ()
-        case _: SEBuiltinRecursiveDefinition => ()
+          if (i > bound) Set(i - bound) else Set() /* adjust to caller's environment */
```
Comment (remyhaemmerle-da) [inline daml-lf/interpreter/src/main/scala/com/digitalasset/daml/lf/speedy/Compiler.scala:1131]:
```suggestion
          if (i > bound) (free + (i - bound)) else free /* adjust to caller's environment */
```

File: daml-lf/interpreter/src/main/scala/com/digitalasset/daml/lf/speedy/Compiler.scala [hunk 3]
```diff
@@ -1125,67 +1125,50 @@ private[lf] final case class Compiler(
...
+        case _: SEBuiltin => Set()
+        case _: SEValue => Set()
+        case _: SEBuiltinRecursiveDefinition => Set()
         case SELocation(_, body) =>
-          go(body)
+          go(body, bound)
         case SEAppGeneral(fun, args) =>
-          go(fun)
-          args.foreach(go)
+          go(fun, bound) ++ args.map(go(_, bound)).reduce(_ ++ _)
```
Comment (remyhaemmerle-da) [inline daml-lf/interpreter/src/main/scala/com/digitalasset/daml/lf/speedy/Compiler.scala:1139]:
```suggestion
              alts.foldLeft(go(scrut, bound, free))((exp, free) => go(exp, bound, free))```

File: daml-lf/interpreter/src/main/scala/com/digitalasset/daml/lf/speedy/Compiler.scala [hunk 4]
```diff
@@ -1125,67 +1125,50 @@ private[lf] final case class Compiler(
...
         case SEAppAtomicFun(fun, args) =>
-          go(fun)
-          args.foreach(go)
+          go(fun, bound) ++ args.map(go(_, bound)).reduce(_ ++ _)
         case SEAppSaturatedBuiltinFun(_, args) =>
-          args.foreach(go)
+          args.map(go(_, bound)).reduce(_ ++ _)
```
Comment (remyhaemmerle-da) [inline daml-lf/interpreter/src/main/scala/com/digitalasset/daml/lf/speedy/Compiler.scala:1143]: args.foldLeft(free)((free, arg) =>  go(arg, bound, free))

File: daml-lf/interpreter/src/main/scala/com/digitalasset/daml/lf/speedy/Compiler.scala [hunk 5]
```diff
@@ -1125,67 +1125,50 @@ private[lf] final case class Compiler(
     * The returned free variables are de bruijn indices
     * adjusted to the stack of the caller. */
   def freeVars(expr: SExpr, initiallyBound: Int): Set[Int] = {
-    var bound = initiallyBound
-    var free = Set.empty[Int]
-
-    def go(expr: SExpr): Unit =
+    def go(expr: SExpr, bound: Int): Set[Int] =
```
Comment (remyhaemmerle-da) [inline daml-lf/interpreter/src/main/scala/com/digitalasset/daml/lf/speedy/Compiler.scala:1128]:
```suggestion
    def go(expr: SExpr, bound: Int, free: Set[Int]): Set[Int] =
```

File: daml-lf/interpreter/src/main/scala/com/digitalasset/daml/lf/speedy/Compiler.scala (hunk 2 above)
Comment (remyhaemmerle-da) [inline daml-lf/interpreter/src/main/scala/com/digitalasset/daml/lf/speedy/Compiler.scala:1131]:
```suggestion
          if (i > bound) (free + (i - bound)) else free /* adjust to caller's environment */
```

File: daml-lf/interpreter/src/main/scala/com/digitalasset/daml/lf/speedy/Compiler.scala [hunk 6]
```diff
@@ -1125,67 +1125,50 @@ private[lf] final case class Compiler(
...
       expr match {
         case SEVar(i) =>
-          if (i > bound)
-            free += i - bound /* adjust to caller's environment */
-        case _: SEVal => ()
-        case _: SEBuiltin => ()
-        case _: SEValue => ()
-        case _: SEBuiltinRecursiveDefinition => ()
+          if (i > bound) Set(i - bound) else Set() /* adjust to caller's environment */
+        case _: SEVal => Set()
```
Comment (remyhaemmerle-da) [inline daml-lf/interpreter/src/main/scala/com/digitalasset/daml/lf/speedy/Compiler.scala:1132]:
```suggestion
        case _: SEVal => free
```

File: daml-lf/interpreter/src/main/scala/com/digitalasset/daml/lf/speedy/Compiler.scala [hunk 7]
```diff
@@ -1125,67 +1125,50 @@ private[lf] final case class Compiler(
...
         case SECase(scrut, alts) =>
-          go(scrut)
-          alts.foreach {
-            case SCaseAlt(pat, body) =>
-              val n = patternNArgs(pat)
-              bound += n; go(body); bound -= n
-          }
+          go(scrut, bound) ++ alts
+            .map { case SCaseAlt(pat, body) => go(body, bound + patternNArgs(pat)) }
+            .reduce(_ ++ _)
```
Comment (remyhaemmerle-da) [inline daml-lf/interpreter/src/main/scala/com/digitalasset/daml/lf/speedy/Compiler.scala:1153]:
```suggestion
            alts.foldLeft(go(scrut, bound, free)) {
            case (free, SCaseAlt(pat, body)) => go(body, bound + patternNArgs(pat), free)
          }
```

File: daml-lf/interpreter/src/main/scala/com/digitalasset/daml/lf/speedy/Compiler.scala [hunk 8]
```diff
@@ -1125,67 +1125,50 @@ private[lf] final case class Compiler(
...
-          go(expr)
+          go(expr, bound)
         case x: SEWronglyTypeContractId =>
           throw CompilationError(s"unexpected SEWronglyTypeContractId: $x")
         case x: SEImportValue =>
           throw CompilationError(s"unexpected SEImportValue: $x")
       }
-    go(expr)
-    free
+    go(expr, initiallyBound)
```
Comment (remyhaemmerle-da) [inline daml-lf/interpreter/src/main/scala/com/digitalasset/daml/lf/speedy/Compiler.scala:1167]:
```suggestion
    go(expr, initiallyBound, Set.empty)
```

File: daml-lf/interpreter/src/main/scala/com/digitalasset/daml/lf/speedy/Compiler.scala (hunk 5 above)
Comment (SamirTalwar-DA) [reply daml-lf/interpreter/src/main/scala/com/digitalasset/daml/lf/speedy/Compiler.scala:1128]: It looks like if you accept all @remyhaemmerle-da's changes, you can add `@tailrec` to this function too.

Comment (remyhaemmerle-da) [reply daml-lf/interpreter/src/main/scala/com/digitalasset/daml/lf/speedy/Compiler.scala:1128]: :+1:

Comment (garyverhaegen-da) [reply daml-lf/interpreter/src/main/scala/com/digitalasset/daml/lf/speedy/Compiler.scala:1128]: It's a tree traversal so I don't think that will be possible. `args.foldLeft(free)((free, arg) =>  go(arg, bound, free))` for example is not a tail call.

Comment (SamirTalwar-DA) [reply daml-lf/interpreter/src/main/scala/com/digitalasset/daml/lf/speedy/Compiler.scala:1128]: Missed that; thanks for explaining.

**Comment by garyverhaegen-da** [comment]
@remyhaemmerle-da I made the change you requested, AFAICT this is now doing exactly one set insertion operation per free variable (and no set union at all).

**Pull request merged**